Contribution Guide (#100)
microsoft/reverse-proxy: contributing.md

* [Active Work](https://github.com/microsoft/reverse-proxy/projects/1) - this board is tracking what YARP team members are working on

Our primary focus is going to be on features and work items that are listed in the planning board for the next release. Every feature should have an issue, where scoping and design will be discussed. If you wish to contribute, we'd prefer to agree to a design in the issue, before submitting a PR. The last thing we want is for you to spend time working on a feature and then have the PR rejected or sit and get stale.

There is a cost to accepting PRs. We really appreciate your help and contributions but it takes time for us to review your code, and the team will be responsible for maintaining it. Our primary focus is going to be on work items listed on the planning board. We're happy to take a look at PRs contributing other features, but our focus will be on the work we already have planned. Those decisions aren't final though, and we change them over time as we learn new things. Feel free to file issues or comment on existing ones if you have new data to provide!

### Extensibility

> One of the primary goals of YARP is to be easily extensible. Each deployemnt situation will be different, and the features that are "in the box" may not do exactly what you need. The goal is for you to be able to insert additional modules in the pipeline, or replace a module to achieve the functionality that you need. The answer to many feature requests may be that it should be a custom module, rather than a change to the existing feature. In those cases, we are unlikely to want a PR for the module, but will be very interested in any changes to the core that enable the extensibility for you to achieve your scenario.

Reviewer: ```suggestion
One of the primary goals of YARP is to be easily extensible. Each deployment situation will be different, and the features that are "in the box" may not do exactly what you need. The goal is for you to be able to insert additional modules in the pipeline, or replace a module to achieve the functionality that you need. The answer to many feature requests may be that it should be a custom module, rather than a change to the existing feature. In those cases, we are unlikely to want a PR for the module, but will be very interested in any changes to the core that enable the extensibility for you to achieve your scenario.
```

> # YARP Contribution Guide

Author: This follows the pattern elsewhere - what's the issue?

> We want YARP to be a true open source project and accept 3rd party contributions, but like all open source projects need to set some guidelines on how and what to contribute. There can be nothing more frustrating than working on a change for some time, and it then sitting forever as a PR. The purpose of this doc is to set expectations for how best to contribute so that YARP can benefit from the communities skills and knowlege.

Reviewer 2: ```suggestion
We're excited to accept contributions, but like all open source projects need to set some guidelines on how and what to contribute. There can be nothing more frustrating than working on a change for some time, only to have it sitting forever as a PR. The purpose of this doc is to set expectations for how best to contribute so that YARP can benefit from the communities skills and knowlege.
```

Let's not separate out "third-party" here. All contributions (even internal ones) should be mostly treated equally. Even if he is on the team, we don't really want @Tratcher dropping a big PR on us that doesn't match our plans ;).

## General feedback and discussions?
Start a discussion on the [repository issue tracker](https://github.com/microsoft/reverse-proxy/issues).

## Issues

> We love issues. Issues are the best place to discuss bugs, feature requests, ideas, designs etc. Issues are the way for everyone to communicate. We use issues to communicate across the team (and not just because we're currently all working from home due to Covid19.) Almost all contributions should start with an issue - even if its just filing the issue.

Reviewer 2: ```suggestion
We love issues. Issues are the best place to discuss bugs, feature requests, ideas, designs etc. Issues are the way for everyone to communicate. We use issues to communicate across the team. Almost all contributions should start with an issue.
```

Stay away from current events, it'll go stale ;).

> If the conversations on an issue to wander off from the initial topic, and new ideas or issues get introduced, then those should be split off into separate issues. That makes it much easier to triage the issue as the decision can/should apply to the main concept, and those separate issues won't be lost and will also be considered. So if in doubt create a new issue, and add links from both the new and original issue to each other. The new issue can always be closed if it turns out to be a duplicate.

Reviewer 2: ```suggestion
If the conversations on an issue wander off from the initial topic, and new ideas or issues get introduced, then those should be split off into separate issues. That makes it much easier to triage the issue as the decision can/should apply to the main concept, and those separate issues won't be lost and will also be considered. So if in doubt create a new issue, and add links from both the new and original issue to each other. The new issue can always be closed if it turns out to be a duplicate.
```

> In particular, conversations on closed issues won't cause an issue to be re-opened, and so are unlikely to be noticed, so please create a new issue and refer to the old one with a link.

Reviewer 2: ```suggestion
In particular, conversations on closed issues *won't cause an issue to be re-opened*, and are unlikely to be noticed, so please create a new issue and refer to the old one with a link.
```

## Bugs

As long as humans write software, there will be bugs. If you find a bug, file an issue. 

> The line between a bug and a feature request or design change are tricky. For the purposes of this section a bug is where the code doesn't do what was intended by the design of the feature - it may be because of human error, or not considering cases that occur in the real world. If you feel confident about the fix, create a Pull Request (PR). Bug PR's should be small and uncontroversial, and therefore easily integrated.

Reviewer 2: ```suggestion
The line between a bug and a feature request or design change are tricky. For the purposes of this section a bug is where the code doesn't do what was intended by the design of the feature - it may be because of human error, or not considering cases that occur in the real world. If you feel confident about the fix, create a Pull Request (PR). Bug PRs should be small and uncontroversial, and therefore easily integrated. The PR should include a test that proves the bug is fixed.
```

nit/pet peeve: apostrophes are for indicating possession, even when acronyms are being used :)

* ["Good first issue" issues](https://github.com/microsoft/reverse-proxy/labels/good%20first%20issue) - we think these are a good for newcomers.

### Identifying the scale

If you would like to contribute to one of our repositories, first identify the scale of what you would like to contribute. If it is small (grammar/spelling or a bug fix) feel free to start working on a fix. If you are submitting a feature or substantial code contribution, please discuss it with the team and ensure it follows the product roadmap. You might also read these two blogs posts on contributing code: [Open Source Contribution Etiquette](http://tirania.org/blog/archive/2010/Dec-31.html) by Miguel de Icaza and [Don't "Push" Your Pull Requests](https://www.igvita.com/2011/12/19/dont-push-your-pull-requests/) by Ilya Grigorik. All code submissions will be rigorously reviewed and tested by the YARP team, and only those that meet an extremely high bar for both quality and design/roadmap appropriateness will be merged into the source.

### Roadmap

We are using project boards to track development of YARP:

> * [Planning](https://github.com/microsoft/reverse-proxy/projects/5) - this board lists the features / work items for the next release of YARP

Reviewer 2: ```suggestion
* [Planning](https://github.com/microsoft/reverse-proxy/projects/5) - this board lists the features / work items for the next major release of YARP
```

* [Active Work](https://github.com/microsoft/reverse-proxy/projects/1) - this board is tracking what YARP team members are working on

> Our primary focus is going to be on features and work items that are listed in the planning board for the next release. At the time of writing we are targetting 1.0, the goal of which is to stand up the core functionality so that people can start trying YARP. Every feature for that mile stone should have an issue, where scoping and design will be discussed. If you wish to contribute, we'd prefer to agree to a design in the issue, before submitting a PR. The last thing we want is for you to spend time working on a feature and then have the PR rejected or sit and get stale.

Reviewer 2: ```suggestion
Our primary focus is going to be on features and work items that are listed in the planning board for the next release. Every feature should have an issue, where scoping and design will be discussed. If you wish to contribute, we'd prefer to agree to a design in the issue, before submitting a PR. The last thing we want is for you to spend time working on a feature and then have the PR rejected or sit and get stale.
```

Remove reference to 1.0 (prevents staleness ;))

> There is a cost to accepting PRs, so our primary focus is going to be on work items listed on the planning board. The decisions about what is on the board was based on the resources of the YARP team, and the knowlege we had at the time. We may be wrong, and the community may have better ideas - create an issue and get some community concensus and we'll reassess our decisions.

Reviewer 2: ```suggestion
There is a cost to accepting PRs. We really appreciate your help and contributions but it takes time for us to review your code, and the team will be responsible for maintaining it. Our primary focus is going to be on work items listed on the planning board. We're happy to take a look at PRs contributing other features, but our focus will be on the work we already have planned. Those decisions aren't final though, and we change them over time as we learn new things. Feel free to file issues or comment on existing ones if you have new data to provide!
```

A suggested rephrase and expanded why PRs aren't free.

### Extensibility

One of the primary goals of YARP is to be easily extensible. Each deployemnt situation will be different, and the features that are "in the box" may not do exactly what you need. The goal is for you to be able to insert additional modules in the pipeline, or replace a module to achieve the functionality that you need. The answer to many feature requests may be that it should be a custom module, rather than a change to the existing feature. In those cases, we are unlikely to want a PR for the module, but will be very interested in any changes to the core that enable the extensibility for you to achieve your scenario.

### Submitting a pull request

You will need to sign a [Contributor License Agreement](https://cla.opensource.microsoft.com) when submitting your pull request. To complete the Contributor License Agreement (CLA), you will need to follow the instructions provided by the CLA bot when you send the pull request. This needs to only be done once for all repos using this CLA.

> If you don't know what a pull request is read this article: https://help.github.com/articles/using-pull-requests. Make sure the repository can build and all tests pass. Familiarize yourself with the project workflow and our coding conventions. We are following the coding, style, and general [engineering guidelines](https://github.com/dotnet/aspnetcore/wiki/Engineering-guidelines) from ASP.NET Core.

Reviewer 2: ```suggestion
If you don't know what a pull request is read this article: https://help.github.com/articles/using-pull-requests. Make sure the repository can build and all tests pass. Familiarize yourself with the project workflow and our coding conventions.
```

I would remove the link, those guidelines are super stale ;)

Author: Is there a common set of guidelines for .NET?

> ### Tests
> 
> -  Tests need to be provided for every bug/feature that is completed.
> -  Tests only need to be present for issues that need to be verified by QA (for example, not tasks)
> -  If there is a scenario that is far too hard to test there does not need to be a test for it.
>   - "Too hard" is determined by the team as a whole.
> 

Reviewer 2: ```suggestion
```

I'm not a fan of this section, suggest yanking it. Some content above already indicated that tests should be present, and these "rules" are full of exceptions, so they don't feel super useful.
